tools/elfdeps.c

```diff
@@ -33,6 +38,109 @@ typedef struct elfInfo_s {
     ARGV_t provides;
 } elfInfo;
 
+/*
+ * If filename contains ".so" followed by a version number, return
+ * a copy of the version number.
+ */
+static char *getLibtoolVer(const char *filename)
+{
+    const char *so;
+    char dest[PATH_MAX];
+    int destsize = 0;
+    int found_digit = 0, found_dot = 0;
+
+    destsize = readlink(filename, dest, PATH_MAX);
+    if (destsize > 0) {
+	dest[destsize] = 0;
+	filename = dest;
+    }
+    /*
+     * Start from the end of the string.  Verify that it ends with
+     * numbers and dots, preceded by ".so.".
+     */
+    so = filename + strlen(filename);
+    while (so > filename+2) {
+	if (*so == '.') {
+	    found_dot++;
+	    so--;
+	    continue;
+	} else if (strchr("0123456789", *so)) {
```

Comment:
You can use `risdigit(*so)` here, this is a locale agnostic helper in rpm and more to the point.

But that made me think of openssl 1.x which have characters appended to the micro version. Openssl has symbol versions so it doesn't matter here, but I wonder if it's something we should handle here. Looking at Fedora repos, characters as the last letter brings up only weird stuff suggesting that maybe not, but dunno:
> $ repoquery --quiet --provides -a|grep "\.so\." |sed -e 's/()(64bit)//g'|grep -v \)$|sort -u|grep -v "[[:digit:]]$"
libbsoncxx.so._noabi
libgnuradio-air_modes.so.1.0.0git
libIndexStore.so.13git
libkicad_3dsg.so.2.0.0-6.0.7-5.fc37.i386.debug
liblldb.so.13git
libLTO.so.13git
libmongocxx.so._noabi
libstdbuf.so.single
libterminal-nautilus.so.

Of those, libgnuradio-air_modes.so.1.0.0git appears a legit library that programs could normally link against.

Reply:
Without defined semantics, it's probably a bad idea to process any non-numeric versions.  Is 13git a version leading up to a stable release of version 13?  If so, that'd be very difficult to correct later:

```
 $ rpmdev-vercmp 13git 13
13git > 13
```

I was less concerned about that class of problems when I suggested using the rpm version, since maintainers already have to make sure that rpm understands the progression of releases.  (Though, on the other hand, I did not initially use the epoch, and that probably would have been necessary.)

Thinking through that further: there would be a problem if libfoo.so.12.0.0 were available in a build root and packages built against it, and later that package were updated with libfoo.so.13.0.0git.  The new version wouldn't "provide" a versioned library, and that'd break dependencies or force the maintainer to manually Provide something.

It bears thinking about further, I guess.  But I do think this is a problem worth solving.  I'll keep working.

Reply:
Breaking dependencies due unexpected junk in the soname is at least something the package maintainer *will* notice, which is far far better than doing something unwanted silently. So limiting to numeric versions is indeed probably the safer bet, and not a show-stopper at all. Just something to think about.